plugins/inputs/win_wmi/win_wmi.go

```diff
@@ -0,0 +1,265 @@
+	// ensure the CoInitialize[Ex]() call is bound to its current OS thread.
+	// Otherwise, attempting to initialize and run parallel queries across
+	// goroutines will result in protected memory errors.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	tags := map[string]string{}
+	fields := map[string]interface{}{}
+
+	// init COM
+	if err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED); err != nil {
+		oleCode := err.(*ole.OleError).Code()
+		if oleCode != ole.S_OK && oleCode != sFalse {
+			return err
+		}
+	}
+	defer ole.CoUninitialize()
+
+	unknown, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
+	if err != nil {
+		return err
+	} else if unknown == nil {
+		return fmt.Errorf("failed to create WbemScripting.SWbemLocator, is WMI broken?")
+	}
+	defer unknown.Release()
+
+	wmi, err := unknown.QueryInterface(ole.IID_IDispatch)
+	if err != nil {
+		return fmt.Errorf("failed to QueryInterface: %v", err)
+	}
+	defer wmi.Release()
+
+	// service is a SWbemServices
+	serviceRaw, err := oleutil.CallMethod(wmi, "ConnectServer", nil, q.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed calling method ConnectServer: %v", err)
+	}
+	service := serviceRaw.ToIDispatch()
+	defer func() { _ = serviceRaw.Clear() }()
+
+	// result is a SWBemObjectSet
+	resultRaw, err := oleutil.CallMethod(service, "ExecQuery", q.query)
+	if err != nil {
+		return fmt.Errorf("failed calling method ExecQuery for query %s: %v", q.query, err)
+	}
+	result := resultRaw.ToIDispatch()
+	defer func() { _ = resultRaw.Clear() }()
+
+	count, err := oleInt64(result, "Count")
+	if err != nil {
+		return fmt.Errorf("failed getting Count: %v", err)
+	}
+
+	for i := int64(0); i < count; i++ {
+		// item is a SWbemObject
+		itemRaw, err := oleutil.CallMethod(result, "ItemIndex", i)
+		if err != nil {
+			return fmt.Errorf("failed calling method ItemIndex: %v", err)
+		}
+
+		outerErr := func() error {
+			item := itemRaw.ToIDispatch()
+			defer item.Release()
+
+			for _, wmiProperty := range q.Properties {
+				innerErr := func() error {
+					prop, err := oleutil.GetProperty(item, wmiProperty)
+					if err != nil {
+						return fmt.Errorf("failed GetProperty: %v", err)
+					}
+					defer func() { _ = prop.Clear() }()
+
+					if q.tagFilter.Match(wmiProperty) {
+						valStr, err := internal.ToString(prop.Value())
+						if err != nil {
+							return fmt.Errorf("converting property %q failed: %v", wmiProperty, err)
+						}
+						tags[wmiProperty] = valStr
+					} else {
+						var fieldValue interface{}
+						switch v := prop.Value().(type) {
+						case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64:
+							fieldValue = v
+						case string:
+							// still might be an int because WMI
+							valInt, err := strconv.ParseInt(v, 10, 64)
+							if err == nil {
+								fieldValue = valInt
+							} else {
+								fieldValue = v
+							}
```

Comment:
This is a common place that bites us badly where data types change because one day it can convert to an int, but the next day it cannot because it overruns or is a float. 

So what about floats? And better yet is it worth leaving it to the user to use a processor to convert these?

Reply:
I wholeheartedly agree and ripped this portion out.

Reply:
I'll note that--after ripping this out--I definitely had to use processors in order to get correct data types with several classes. Accordingly, I added a note in the troubleshooting section of the documentation under a "data types" subheading.